clients/drcachesim/tracer/syscall_pt_trace.cpp

```diff
@@ -0,0 +1,144 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* syscall_pt_trace.cpp: module for recording kernel PT traces for every syscall. */
+
+#include <cstring>
+
+#include "dr_api.h"
+#include "drpttracer.h"
+#include "syscall_pt_trace.h"
+
+#ifndef BUILD_PT_TRACER
+#    error "This module requires the drpttracer extension."
+#endif
+
+#define PT_DATA_FILE_NAME_SUFFIX ".pt"
+#define PT_METADATA_FILE_NAME_SUFFIX ".metadata"
+
+syscall_pt_trace_t::syscall_pt_trace_t()
+    : write_file_func_(nullptr)
+    , pttracer_handle_ {}
+    , recorded_syscall_num_(0)
+    , cur_recording_sysnum_(-1)
+    , drcontext_(nullptr)
+    , pt_dir_name_ { '\0' }
+{
+}
+
+syscall_pt_trace_t::~syscall_pt_trace_t()
+{
+}
+
+bool
+syscall_pt_trace_t::init(void *drcontext, char *pt_dir_name, size_t pt_dir_name_size,
+                         ssize_t (*write_file_func)(file_t file, const void *data,
+                                                    size_t count))
+{
+    drcontext_ = drcontext;
+    memcpy(pt_dir_name_, pt_dir_name, pt_dir_name_size);
+    write_file_func_ = write_file_func;
+    return true;
+}
+
+bool
+syscall_pt_trace_t::start_syscall_pt_trace(int sysnum)
+{
+#define RING_BUFFER_SIZE_SHIFT 8
+    ASSERT(drcontext_ != nullptr, "drcontext_ is nullptr");
+    ASSERT(pttracer_handle_.handle == nullptr, "pttracer_handle_.handle isn't nullptr");
+    if (drpttracer_create_handle(drcontext_, DRPTTRACER_TRACING_ONLY_KERNEL,
+                                 RING_BUFFER_SIZE_SHIFT, RING_BUFFER_SIZE_SHIFT,
+                                 &pttracer_handle_.handle) != DRPTTRACER_SUCCESS) {
+        return false;
+    }
+    if (drpttracer_start_tracing(drcontext_, pttracer_handle_.handle) !=
+        DRPTTRACER_SUCCESS) {
+        return false;
+    }
+    cur_recording_sysnum_ = sysnum;
+    return true;
+}
+
+bool
+syscall_pt_trace_t::stop_syscall_pt_trace()
+{
+    ASSERT(drcontext_ != nullptr, "drcontext_ is nullptr");
+    ASSERT(pttracer_handle_.handle != nullptr, "pttracer_handle_.handle is nullptr");
+
+    drpttracer_output_cleanup_last_t output;
+    if (drpttracer_stop_tracing(drcontext_, pttracer_handle_.handle, &output.data) !=
+        DRPTTRACER_SUCCESS) {
+        return false;
+    }
+    pttracer_handle_.~drpttracer_handle_cleanup_last_t();
```

Comment:
This destroys the handle: but the .dox file says the same handle can be used across multiple sessions.  So one of these is incorrect.

Reply:
Curretly, using drpttracer_start_tracing() and drpttracer_stop_tracing() will generate a PT fragment, which may not contains PSB packets. So drpt2ir can't to decode the fragment successful. In future, if I upgrade drpt2ir to support decoding the PT fragment, I will update this part.